feature-pitches/002-file-based-workflows/002-file-based-workflows.md

```diff
@@ -0,0 +1,109 @@
+
+The consequence of this is that there is no connection between iterations of a single workflow file as the user develops it. We recognize that this not *ideal* (for example, the user will only be able to compare `activities` generated by a non-modified `plan`; once they change the workflow, activities are not easily comparable anymore), however we consider it a worthwhile tradeoff for the sake of making workflow development an easy user experience.
+
+
+**How do we meaningfully list *relevant* workflow metadata with so many pieces of workflow metadata floating around?**
+
+The one cost of creating lots of workflow metadata is that listing the project's workflows yields a lot of metadata that the user doesn't care about. But the solution to that is right in the statement- the user doesn't care about any metadata that relate to data objects that no longer exist. `renku workflow ls` should only list workflows that relate to currently existing code and data objects, i.e. show me the workflow that I ran to create the current version of file `b`, not the workflow I ran yesterday that created a previous version of file `b`. (In the `plan` and `activity` vocabulary, only `plans` that have an `activity` that relates to a file currently in the project)
+
+
+### Misc Rabbit Holes
+
+**How do you handle when a workflow step generates many files?**
+
+As is the state of the current workflow functionality, inputs and outputs may be directories, rather than files, to support multiple files. However we don’t support any kind of wildcard syntax.
+
+**How does this feature mesh with the current workflow parameter file functionality?**
+
+Currently, the user has the option of creating a file specifying workflow parameters, and passing this in to `renku workflow execute` to override the default parameters. In the future, the combination of this parameter file and the workflow definition file could develop into a sort of base workflow + several “variation” files defining a set of related workflow experiments. But work into this direction is _out of scope_ of this pitch. 
+
+
+**Should one workflow file be able to reference another workflow file? (i.e. to compose workflows between files)**
+
+My default is *no*, keep it simple for the scope of this pitch.
+
+**Potential for Un-executable `Plans`**
+
+In the current `renku run` implementation, a `plan` is only created after the command executed successfully. By importing and creating the `plan` upon executing the workflow file, the user could potentially end up with a non-executable `plan` in their project. This sounds non-ideal, however, it doesn’t really have any implications for the user, so it’s up to the team for how they would like to handle this behavior.
```

Comment:
The related problem I see here is the issue of _when_ one of the workflow files imagined in this pitch gets processed, i.e. added to the metadata of the project. Does it get processed when the file is committed via some sort of hook? Or does it not exist in the project's metadata until it is used? Both options are not ideal - hooks are inherently imperfect (the user simply might not have them installed) and if we wait for the workflow to actually be used before it "exists", we automatically invalidate any "worklow-in-progress" concept that might be natural to have e.g. in the UI.

Reply:
Could you elaborate on the "workflow-in-progress" concept? I'm not sure what you mean.

Reply:
I think it's moot given that somewhere else it was decided that the workflow needs to be executed before it gets added to the project's metadata. Until that point it's basically invisible in any workflow front-end.